lib/render.js

```diff
@@ -25,28 +27,39 @@ function composeData(data) {
 }
 
 /**
- * Take in the data returned from Amphora and turn it into HTML output.
+ * Take in the data returned from Amphora, call `render` hooks,
+ * and return HTML output.
  *
  * @param  {Object} data
- * @return {Promise}
+ * @return {Promise<String>} html
  */
 function render(data) {
-  const template = clayUtils.getComponentName(data._layoutRef || data._self),
-    rootPartial = hbs.partials[template],
-    composedData = composeData(data);
-
-  if (!rootPartial) {
-    throw new Error(`Missing template for ${data.template}`);
-  }
-
-  return Promise.resolve(rootPartial(composedData))
-    .then(mediaService.append(data._media, data.locals.edit, data.locals.site))
-    .then(result => {
-      return {
-        output: result,
-        type: 'html'
-      };
-    });
+
+  const hooks = _.chain(setup.plugins)
+                  .map('render')
+                  .compact()
+                  .value();
+
+  return bluebird.reduce(hooks, (data, hook) => hook(data), data)
```

Comment:
I'm initially scared of passing in the full `data` object here. I don't know that the API should allow for the mutation of `_version`,  `_componentSchemas`, etc. Since other packages (namely Kiln) rely on some of these values, I'd prefer not to give plugins too many capabilities.

What I would propose is passing through exactly what `model.js`/`upgrade.js`/new pub service receive: `uri`, `data`, `locals`.

But even then, I'd apply some limitations there. I don't know if plugins should be able to mutate locals, only read from the object. I'm up for debate on this point, but I do think the function signature should remain the same as so many other parts of Clay: get these three arguments and return some data.

Reply:
I like the idea of sticking with the `uri, data, locals` signature.

For `locals`, we could deep copy it or `Object.freeze` it, but either one would have a perf hit, so I vote for just warning against mutating in the README.md.